gufe/base.py

```diff
@@ -0,0 +1,229 @@
+import abc
+import uuid
+import sys
+import threading
+import hashlib
+import inspect
+from copy import copy
+from packaging.version import parse as parse_version
+from typing import Dict, Any, Callable
+
+
+TOKENIZABLE_CLASS_REGISTRY = {}
+
+def register_tokenizable_class(cls):
+    TOKENIZABLE_CLASS_REGISTRY[cls.__name__] = cls
+
+    
+class _GufeTokenizableMeta(type):
+    def __new__(cls, name, bases, classdict):
+        componentcls = super().__new__(cls, name, bases, classdict)
+        register_tokenizable_class(componentcls)
+        return componentcls
+
+
+class _ABCGufeClassMeta(_GufeTokenizableMeta, abc.ABCMeta):
+    # required to make use of abc.ABC in classes that use _ComponentMeta metaclass
+    # see https://stackoverflow.com/questions/57349105/python-abc-inheritance-with-specified-metaclass 
+    ...
+        
+
+class GufeTokenizable(metaclass=_ABCGufeClassMeta):
+    """Base class for all tokenizeable gufe objects.
+
+    """
+    def __init__(self):
+
+        # since __hash__ isn't inheritable, monkey-patching is the best we can
+        # do besides defining it in every subclass
+        self.__class__.__hash__ = GufeTokenizable.__hash__
+
+    def __lt__(self, other):
+        return hash(self) < hash(other)
+
+    def __eq__(self, other):
+        if not isinstance(other, self.__class__):
+            return NotImplemented("Component comparisons must be of the same type")
+
+        return self.to_dict() == other.to_dict()
+
+    def __hash__(self):
+        return hash(self.key)
+
+    @property
+    def token(self):
+        if not hasattr(self, '_token') or self._token is None:
+            self._token = tokenize(self)
+        return self._token
+
+    @property
+    def key(self):
+        if not hasattr(self, '_key') or self._key is None:
+            prefix = type(self).__name__
+            self._key = GufeKey(f"{prefix}-{self.token}")
+
+        # add to registry if not already present
+        TOKENIZABLE_REGISTRY[self._key] = self
+
+        return self._key
+
+    @property
+    def defaults(self):
+        sig = inspect.signature(self.__init__)
+
+        defaults = {
+            param.name: param.default for param in sig.parameters.values()
+            if param.default is not inspect.Parameter.empty
+        }
+
+        return defaults
+
+    # TODO: rewrite the below three functions to be fully recursive like 
+    # _dictdecode_dependencies
+    def _keyencode_dependencies(self, d):
+        for key, value in d.items():
+            if isinstance(value, dict):
+                self._keyencode_dependencies(value)
+            elif isinstance(value, list):
+                for i, item in enumerate(value):
+                    if hasattr(item, '_gufe_tokenize'):
+                        d[key][i] = item.key
+            else:
+                if hasattr(value, '_gufe_tokenize'):
+                    d[key] = value.key
```

Comment:
I think this instead needs to be a dict that includes the key, perhaps something as simple as `{":gufe-key:": value.key}`. When loading from storage, we need some way of recognizing that this is "one of ours". It won't already be in the registry, so it needs to be something intrinsic to the serialized format.

Reply:
ah, got confused between token and key here. Key is fine; token is not

Reply:
Agreed, we'll proceed with {':gufe-key:': str(<GufeKey>)} as you suggest! I don't see any issue with this choice of structure.